### PR TITLE
KAFKA-3598: Improve JavaDoc of public API

### PR DESCRIPTION
**Pull request by mjsax**

### CODE REVIEW

**Comment by mjsax** [comment]
@guozhangwang Please review. There are still some TODOs that need to be removed before this can get merged. Please comment on TODOs.

File: config/server.properties [hunk 1]
```diff
@@ -114,4 +114,4 @@ zookeeper.connect=localhost:2181
 # Timeout in ms for connecting to zookeeper
 zookeeper.connection.timeout.ms=6000
 
-
+delete.topic.enable=true
```
Comment (miguno) [inline config/server.properties]: Why was this added to server.properties?  This looks like a mistake to me.

File: streams/src/main/java/org/apache/kafka/streams/KeyValue.java [hunk 2]
```diff
@@ -29,14 +29,31 @@
...
     }
 
+    /**
+     * Create a new key-value pair.
+     *
+     * @param key    the key
+     * @param value  the value
+     * @param <K>    the type of the key
+     * @param <V>    the type of the value
+     * @return       a new typed key value pair
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/KeyValue.java]: Nitpicks:
- Why "typed" key value pair?  Of course it is typed, no?
- To be consistent we should write "key-value pair", not "key value pair".

File: streams/src/main/java/org/apache/kafka/streams/processor/WallclockTimestampExtractor.java [hunk 3]
```diff
@@ -28,6 +28,13 @@
  * built-in <i>CreateTime</i> timestamp (see KIP-32: Add timestamps to Kafka message for details).
  */
 public class WallclockTimestampExtractor implements TimestampExtractor {
+
+    /**
+     * Return the current wall clock time as timestamp.
+     *
+     * @param record  a message
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/processor/WallclockTimestampExtractor.java]: Note: In the Streams terminology, we say "record" (or "data record").  We don't use the term "message" (we say "message" in Kafka).  This terminology change needs to be done throughout the PR.

File: streams/src/main/java/org/apache/kafka/streams/processor/WallclockTimestampExtractor.java [hunk 4]
```diff
@@ -28,6 +28,13 @@
...
+     * @return        the current wall clock time
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/processor/WallclockTimestampExtractor.java]: "the current wall clock time, expressed in milliseconds since midnight, January 1, 1970 UTC"

File: streams/src/main/java/org/apache/kafka/streams/kstream/Aggregator.java [hunk 5]
```diff
@@ -18,13 +18,22 @@
...
  *
  * @param <K>   key type
  * @param <V>   original value type
  * @param <T>   aggregate value type
  */
 public interface Aggregator<K, V, T> {
 
+    /**
+     * Compute a new aggregate for the message with the given key and value and the current aggregate (ie,
+     * {@link #apply(Object, Object, Object)} "increments" an aggregate with another message).
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Aggregator.java]: I don't understand:

> (ie, {@link #apply(Object, Object, Object)} "increments" an aggregate with another message)

Where does this example imply that the function is incrementing?

Also, typo: "ie" -> "i.e."

File: streams/src/main/java/org/apache/kafka/streams/kstream/HoppingWindows.java [hunk 6]
```diff
@@ -23,14 +23,16 @@
...
+ * The hopping window specifications used for aggregations. TODO extend
  */
 public class HoppingWindows extends Windows<HoppingWindow> {
 
     private static final long DEFAULT_SIZE_MS = 1000L;
 
+    /** The size of the window in milliseconds. */
     public final long size;
 
+    /** The period (ie, advance) of the window in milliseconds. */
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/HoppingWindows.java]: "ie" -> "i.e."

File: streams/src/main/java/org/apache/kafka/streams/kstream/HoppingWindows.java [hunk 7]
```diff
@@ -23,14 +23,16 @@
...
 /**
- * The hopping window specifications used for aggregations.
+ * The hopping window specifications used for aggregations. TODO extend
  */
 public class HoppingWindows extends Windows<HoppingWindow> {
 
     private static final long DEFAULT_SIZE_MS = 1000L;
 
+    /** The size of the window in milliseconds. */
     public final long size;
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/HoppingWindows.java]: @guozhangwang: Slightly off-topic question:  Should `size` and `period` be renamed to `sizeMs` and `periodMs`, respectively?

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java [hunk 8]
```diff
@@ -23,11 +23,13 @@
 import java.util.Map;
 
 /**
- * The window specifications used for joins.
+ * The window specifications used for joins. TODO extend
  */
 public class JoinWindows extends Windows<TumblingWindow> {
 
+    /** Maximum time difference for tuples that are before the join tuple. */
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java]: What's the time unit?  Millseconds?

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java [hunk 9]
```diff
@@ -23,11 +23,13 @@
...
     public final long before;
+    /** Maximum time difference for tuples that are after the join tuple. */
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java]: What's the time unit?  Millseconds?

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java [hunk 10]
```diff
@@ -23,11 +23,13 @@
 import java.util.Map;
 
 /**
- * The window specifications used for joins.
+ * The window specifications used for joins. TODO extend
  */
 public class JoinWindows extends Windows<TumblingWindow> {
 
+    /** Maximum time difference for tuples that are before the join tuple. */
     public final long before;
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java]: @guozhangwang: See previous question -- rename these fields to `beforeMs` and `afterMs`?

File: streams/src/main/java/org/apache/kafka/streams/kstream/HoppingWindows.java [hunk 11]
```diff
@@ -23,14 +23,16 @@
 import java.util.Map;
 
 /**
- * The hopping window specifications used for aggregations.
+ * The hopping window specifications used for aggregations. TODO extend
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/HoppingWindows.java]: I would not add all these `TODO` entries _within_ the public java docs.

Rather, I'd add the entry after the docstring:

``` java
/**
 * The hopping window [...]
*/
// TODO: extend the documentation
public class ...
```

File: streams/src/main/java/org/apache/kafka/streams/kstream/UnlimitedWindows.java [hunk 12]
```diff
@@ -38,12 +39,18 @@ private UnlimitedWindows(String name, long start) {
...
     /**
-     * Returns an unlimited window definition
+     * Return an unlimited window definition TODO ?
      */
     public static UnlimitedWindows of(String name) {
         return new UnlimitedWindows(name, DEFAULT_START_TIMESTAMP);
     }
 
+    /**
+     * Return a new unlimited window for the specified start timestamp. TODO ?
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/UnlimitedWindows.java]: Again, we should not track `TODO` within the public docstring.

File: streams/src/main/java/org/apache/kafka/streams/kstream/Window.java [hunk 13]
```diff
@@ -18,36 +18,56 @@
...
  */
 public abstract class Window {
 
     private long start;
     private long end;
 
+    /**
+     * Create a new window for the given start time (inclusive) and end time (exclusive).
+     *
+     * @param start  the start timestamp of the window (inclusive)
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Window.java:31]: We should also always document the relevant time unit (here: milliseconds) throughout this PR.

File: streams/src/main/java/org/apache/kafka/streams/kstream/Window.java [hunk 14]
```diff
@@ -18,36 +18,56 @@
...
+     *
+     * @param other  another window
+     * @return       {@code true} if {@code other} overlaps with this window&mdash;{@code false} otherwise
+     */
     public boolean overlap(Window other) {
         return this.start() < other.end() || other.start() < this.end();
     }
 
+    /**
+     * Check if the start and end timestamp of the given window are equal to the start and end timestamp
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Window.java]: Typo: "Check if the start and end timestamp[s] of the given window are equal to the start and end timestamp[s] of this window[, respectively]."

File: streams/src/main/java/org/apache/kafka/streams/kstream/Window.java [hunk 15]
```diff
@@ -18,36 +18,56 @@
...
+     * of this window.
+     * 
+     * @param other  another window
+     * @return       {@code true} if start and end timestamp of {@code other} are equal to start end timestamp of this
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Window.java]: See typo fix above.

Nitpick: Also, why the `&mdash;`?  Typically you'd use a comma or semicolon (e.g. [Boolean.equals()](https://docs.oracle.com/javase/7/docs/api/java/lang/Boolean.html#equals%28java.lang.Object%29)):

> True if ...; false otherwise.

File: streams/src/main/java/org/apache/kafka/streams/kstream/Windowed.java [hunk 16]
```diff
@@ -26,26 +26,36 @@
  */
 public class Windowed<T> {
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Windowed.java]: @guozhangwang : This class explicitly talks about being "The windowed **key** interface used".  For consistency, should we then also replace `T` with `K` to highlight the focus on keys?  Alternatively, we could also keep `Windowed<T>` and simply remove the "key" focus from the docs -- it's only because we use `Windowed` for only keys in Streams, but there's nothing key-specific in this implementation.

File: streams/src/main/java/org/apache/kafka/streams/kstream/Windowed.java [hunk 17]
```diff
@@ -26,26 +26,36 @@
...
 
-    private T value;
+    private T key;
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Windowed.java]: Good catch!

File: streams/src/main/java/org/apache/kafka/streams/kstream/Windows.java [hunk 18]
```diff
@@ -83,19 +83,24 @@ protected Windows segments(int segments) {
         return this;
     }
 
+    // TODO?
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Windows.java]: Shouldn't we also fix these while we're at it?

File: streams/src/main/java/org/apache/kafka/streams/kstream/Windowed.java [hunk 19]
```diff
@@ -26,26 +26,36 @@
...
     }
 
-    public T value() {
-        return value;
+    /**
+     * Return the key of the window.
+     *
+     * @return the key of the window
+     */
+    public T key() {
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Windowed.java]: Should this method be called `get()` instead (similar to the upcoming `Bytes#get()`)? /cc @guozhangwang 

There's nothing key-specific in this implementation.

File: streams/src/main/java/org/apache/kafka/streams/processor/TaskId.java [hunk 20]
```diff
@@ -25,11 +25,13 @@
 import java.nio.ByteBuffer;
 
 /**
- * The task id representation composed as topic group id plus the assigned partition id.
+ * The task ID representation composed as topic group ID plus the assigned partition ID.
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/processor/TaskId.java:28]: Nitpick:

> Represents the task id as a composition of topic group id and the assigned partition id.

(This change also implies that we do not commit on a contract of how this composition is implemented, i.e. we leave it open whether it's topic-group-id followed by assigned partition id, or vice versa, or...)

File: streams/src/main/java/org/apache/kafka/streams/processor/TimestampExtractor.java [hunk 21]
```diff
@@ -26,10 +26,10 @@
 public interface TimestampExtractor {
 
     /**
-     * Extracts a timestamp from a message
+     * Extracts a timestamp from a message.
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/processor/TimestampExtractor.java]: Should be sth like:

> Extracts a timestamp from a record.
> 
> Typically, the timestamp should represent the milliseconds since midnight, January 1, 1970 UTC.

@guozhangwang : Should we say "should" or "must" represent the milliseconds since the epoch?  KIP-32, for example, explicitly defines the contract to be milliseconds since the epoch, though in the context Streams there's nothing that would specifically force the user to stick to that definition -- a custom timestamp extractor could, for example, simple return seconds since 1 April 2000, and the Streams app would still work (as long as you are not joining such a stream with a stream that has a different interpretation/semantically different extraction of time).

File: streams/src/main/java/org/apache/kafka/streams/processor/WallclockTimestampExtractor.java [hunk 22]
```diff
@@ -28,6 +28,13 @@
  * built-in <i>CreateTime</i> timestamp (see KIP-32: Add timestamps to Kafka message for details).
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/processor/WallclockTimestampExtractor.java]: @guozhangwang: After talking to @ijuma I realized that `ConsumerRecordTimestampExtractor` (to which we refer here) is actually not always returning the CreateTime timestamp (= producer time) -- it could also be LogAppendTime (broker time), and which of the two you'd get depends on the topic-level configuration, right?

If we do need to fix this, we'd have to update the reference in `WallclockTimestampExtractor` and also update `ConsumerRecordTimestampExtractor` itself.

File: streams/src/main/java/org/apache/kafka/streams/state/StateSerdes.java [hunk 23]
```diff
@@ -63,46 +73,105 @@ public StateSerdes(String stateName,
         this.valueSerde = valueSerde;
     }
 
+    /**
+     * Return the key serdes.
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/state/StateSerdes.java]: Typo: "serdes" -> "serde" (singular, not plural).  Same for the similar lines below.

File: streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformerSupplier.java [hunk 24]
```diff
@@ -18,9 +18,14 @@
 package org.apache.kafka.streams.kstream;
 
 /**
- * A value transformer supplier which can create one or more {@link ValueTransformer} instances.
+ * A {@link ValueTransformerSupplier} interface which can create one or more {@link ValueTransformer} instances.
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformerSupplier.java:21]: Question (/cc @guozhangwang):  What determines/influences whether it would create one instance or multiple instanes of `ValueTransformer`?  Is it about whether calling `get()` multiple times would always return the same `ValueTransformer` instance vs. new instances on each invocation?  I am asking because there's no parameter to `get()` that would allow the user to make context-aware decisions on which specific instance/behavior of `ValueTransformer` would need to be returned.

File: streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformer.java [hunk 25]
```diff
@@ -20,7 +20,7 @@
 import org.apache.kafka.streams.processor.ProcessorContext;
 
 /**
- * A stateful Value Transformer interface for transform a value into a new value.
+ * A stateful {@link ValueTransformer} interface for transform a value into a new value.
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformer.java]: Typo: "for transform" -> "to transform"

Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformer.java]: What about:

> Defines a stateful transformation of a record value into a new value.

(Not sure whether this wording approach is consistent with other interface docstrings.)

File: streams/src/main/java/org/apache/kafka/streams/kstream/ValueMapper.java [hunk 26]
```diff
@@ -18,12 +18,18 @@
 package org.apache.kafka.streams.kstream;
 
 /**
- * The KeyValueMapper interface for mapping an original value to a new value (could be another key-value pair).
+ * The {@link KeyValueMapper} interface for mapping an original value to a new value (could be another key-value pair).
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/ValueMapper.java]: Typo: "KeyValueMapper" -> "ValueMapper".

Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/ValueMapper.java]: What about:

> Defines a stateless transformation of a record value into a new value.

(Not sure whether this wording approach is consistent with other interface docstrings.)

File: streams/src/main/java/org/apache/kafka/streams/kstream/ValueJoiner.java [hunk 27]
```diff
@@ -18,13 +18,20 @@
 package org.apache.kafka.streams.kstream;
 
 /**
- * The ValueJoiner interface for joining two values and return a the joined new value.
+ * The {@link ValueJoiner} interface for joining two values into a new value.
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/ValueJoiner.java:21]: What about:

> Defines a stateless transformation of two record values into a new value.

(Not sure whether this wording approach is consistent with other interface docstrings.)

File: streams/src/main/java/org/apache/kafka/streams/kstream/TumblingWindows.java [hunk 28]
```diff
@@ -23,12 +23,13 @@
 import java.util.Map;
 
 /**
- * The tumbling window specifications used for aggregations.
+ * The tumbling window specifications used for aggregations. TODO extend
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/TumblingWindows.java]: What about:

> Represents a tumbling window, which may be used for windowed aggregations.
> 
> The semantics of a tumbling window are: every T (`size`) milliseconds, compute the aggregate total for the same T milliseconds.  Tumbling windows discretize a stream into non-overlapping windows, which implies that a record is only ever contained in one and only one tumbling window.  Tumbling windows are a specialization of {@link HoppingWindows}.

File: streams/src/main/java/org/apache/kafka/streams/kstream/UnlimitedWindows.java [hunk 29]
```diff
@@ -38,12 +39,18 @@ private UnlimitedWindows(String name, long start) {
     }
 
     /**
-     * Returns an unlimited window definition
+     * Return an unlimited window definition TODO ?
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/UnlimitedWindows.java]: This sounds confusing (both the original and the updated docstring) -- "return an unlimited window definition"?

File: streams/src/main/java/org/apache/kafka/streams/kstream/HoppingWindows.java (hunk 11 above)
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/HoppingWindows.java]: What about the following docstring (for starters):

> Represents a hopping window, which may be used for windowed aggregations.
> 
> The semantics of a hopping window are: Every T1 (`period`) milliseconds, compute the aggregate total for T2 (`size`) milliseconds.  Hopping windows are a generalization of {@link TumblingWindows} and discretize a stream into overlapping windows (unless `period == size`, in which case the hopping window is effectively a tumbling window of the same `size`), which implies that a record maybe contained in one and or more "adjacent" sliding windows.

File: streams/src/main/java/org/apache/kafka/streams/kstream/TumblingWindows.java (hunk 28 above)
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/TumblingWindows.java]: We may also need to explain the roles/impacts of (a) the emit behavior and, closely related, (b) the semantics of the timestamp extractor:  Think: "emits are always triggered based on wall clock time, but windows -- and how records are assigned to windows -- are always based on the configured timestamp extractor, which means it could be wall clock time, producer time, or ...

Also, we are always emitting data downstream for every incoming data record ("the KTable as a changelog stream"), i.e. we are not just emitting the "final" aggregate value at the end of a window, right?  So we may need to be a bit more detailed which conditions/events would trigger an emit of data to downstream processor nodes (not sure whether this info would need to be added to this docstring though).

Same for the docstring of `HoppingWindows`.

File: streams/src/main/java/org/apache/kafka/streams/kstream/Aggregator.java (hunk 5 above)
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Aggregator.java]: From my understanding, `apply` is called for each record to update the current (intermediate) agg-value. Something like this (assume the aggregation is a sum for input `3,2,8` and initial agg-value `0`)

`apply(x, 3, 0) => 3`
`apply(x, 2, 3) => 5`
`apply(x, 8, 5) => 11`

Thus, the sum in incrementally build over all calls to `apply()`. In my understanding, increment does not mean `+1`, but has a more general meaning "next step of the aggregation" -- I also put it into quotes as a try to make this clear. But my understanding might not be common sense, and thus, "increments" might be miss leading.

I thought it might be good to explain the user, that the computed/returned agg-result will be handed in as a parameter in the next call -- thus I put this additional explanation. If you think it is more miss leasing than helpful, we should just remove it. Or maybe you have a better idea to phrase this? (Or is my understanding of `apply()` wrong?)

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java (hunk 8 above)
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java]: Never mind my last comment. `TimestampExtractor.extract()` requires the timestamp to be in milliseconds. Thus, it is ms here, too. Will update it.

File: streams/src/main/java/org/apache/kafka/streams/kstream/HoppingWindows.java (hunk 11 above)
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/HoppingWindows.java]: The TODO was not meant to get merged. I was just a tag for this PR to get the reviewers attention, that I don't know how to improve this. I will use your suggestion :)

File: streams/src/main/java/org/apache/kafka/streams/kstream/Window.java (hunk 15 above)
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Window.java]: Personal style only. It is also used in `KStream` and `KTable`. I don't care about it too much. We should agree on a unique style thought. Maybe you can just discuss this shortly and let me know.

File: streams/src/main/java/org/apache/kafka/streams/kstream/Windows.java (hunk 18 above)
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Windows.java]: Yes. I don't understand the semantics of those methods... Need your input.

File: streams/src/main/java/org/apache/kafka/streams/processor/TaskId.java (hunk 20 above)
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/TaskId.java:28]: According to the implementation is it `group-id-partition-id`. But why would this be important?

**Comment by mjsax** [comment]
Thanks for the review @miguno I wait for your and @guozhangwang comments to open points before pushing a new version.

File: streams/src/main/java/org/apache/kafka/streams/kstream/Aggregator.java (hunk 5 above)
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Aggregator.java]: What about:

``` java
    /**
     * Computes a new aggregate from the key and value of a record
     * and the current aggregate for the same key.
     *
     * @param aggKey     the key of the record
     * @param value      the value of the record
     * @param aggregate  the current aggregate
     * @return           the new aggregate value
     */
    T apply(K aggKey, V value, T aggregate);
```

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java (hunk 8 above)
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java]: Actually, having thought more about it, there's nothing in Kafka Streams that would enforce/require milliseconds or milliseconds-since-the-epoch semantics.  (Note that, for the embedded timestamps introduced in KIP-32, the contract of Kafka is to have these timestamps represent milliseconds-since-the-epoch).

For example, a user could embed minute-based timestamps into the payload (message value) of her Kafka messages, then define a custom timestamp extractor that would return these minute-based timestamps as-is (i.e. no need to multiply by `60 * 1000` to convert to milliseconds), and configure any subsequent window sizes etc. with numbers representing minutes -- and it would work just fine.  Correct, @guozhangwang ?

Btw, this "open" approach might be another reason why we should consider introducing `TimeUnit` based configuration of these parameters, e.g. `Windows.of(...).size(3, TimeUnit.SECONDS)`.

File: streams/src/main/java/org/apache/kafka/streams/processor/TaskId.java (hunk 20 above)
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/processor/TaskId.java:28]: Oh, the point I was trying to make: The actual implementation is, as you said, not important (and not part of the contract), thus we should not use wording that might suggest it is.  That said, I realize I might be overthinking this.  Feel free to keep the PR code as is.

File: streams/src/main/java/org/apache/kafka/streams/kstream/Window.java (hunk 15 above)
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Window.java]: I git-grep'ed the source, and (as you said), we use `mdash` elsewhere.  So let's stick to that.

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java (hunk 8 above)
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java]: +1 from my side for all of this.

File: streams/src/main/java/org/apache/kafka/streams/processor/WallclockTimestampExtractor.java (hunk 3 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/WallclockTimestampExtractor.java]: +1

File: streams/src/main/java/org/apache/kafka/streams/kstream/HoppingWindows.java (hunk 7 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/HoppingWindows.java]: The size / period are defined on event time, and hence not necessarily Ms, we should instead remove default values and any `ms` references in those specs. The removal is already done in your PR.

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java (hunk 8 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java]: Yes, the time units in those windowing specs should not be defined by the framework, as they are based on event time.

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java (hunk 9 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java]: Ditto above.

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java (hunk 10 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java]: Ditto above; we should remove `ms` from all windowing specs.

File: streams/src/main/java/org/apache/kafka/streams/kstream/Window.java (hunk 13 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Window.java:31]: Ditto above, no millisecond definition throughout the windowing documentation.

File: streams/src/main/java/org/apache/kafka/streams/kstream/Windowed.java (hunk 16 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Windowed.java]: +1 for changing to `K`.

File: streams/src/main/java/org/apache/kafka/streams/kstream/Windows.java (hunk 18 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Windows.java]: We should remove the emit calls in all the windowing specs for 0.10.0.0 as it is not implemented yet. @mjsax Do you want to do that in this PR, or me / @miguno can do that in a separate PR.

File: streams/src/main/java/org/apache/kafka/streams/kstream/Windowed.java (hunk 19 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Windowed.java]: I think we can just use `key()` as we define this class is only for windowed key.

File: streams/src/main/java/org/apache/kafka/streams/processor/TimestampExtractor.java (hunk 21 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/TimestampExtractor.java]: Discussed offline, for now we do not enforce any semantics in the `TimestampExtractor`. We may want to expand this in the future though, so that this function returns a `Timestamp` object which contains the `long` value as well as the `TimeUnit`, and then in the DSL users will not need to do the manual conversion but just do `(5, TimeUnit.MINUTE)`.

File: streams/src/main/java/org/apache/kafka/streams/processor/WallclockTimestampExtractor.java (hunk 22 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/WallclockTimestampExtractor.java]: Agreed.

File: streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformerSupplier.java (hunk 24 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformerSupplier.java:21]: Always new instance on each invocation. This is because multiple instances may need to have its own local variables and we cannot share a single instance across multiple tasks / threads.

**Comment by guozhangwang** [comment]
Went through the open questions @miguno raised.

**Comment by mjsax** [comment]
@guozhangwang @miguno @enothereska Just updated this.

I just stepped through this again. What about `StreamsMetrics`, `StateRestoreCallback` and `ProcessorContext` interface? Should we extend the JavaDoc for them?

I reverted changes in `TumblindWindow` and `HoppingWindow` to avoid merge conflicts -> will be `TimeWindow` after #1277 got merged.

Should we try to get this two into the release, too? I did not address both in this PR to avoid conflicts.
- https://issues.apache.org/jira/browse/KAFKA-3440
- https://issues.apache.org/jira/browse/KAFKA-3574

**Comment by guozhangwang** [comment]
@mjsax I can merge in #1277 first so that you can save the effort of javadocs for dropped classes. It will be great to extend the javadoc in those three classes as well; in order to not drag too long for this PR, let's do KAFKA-3440 / 3574 in separate PRs.

**Comment by mjsax** [comment]
@guozhangwang Ok. Whatever merge order is better. I will assign KAKFA-3440 und KAFKA-3574 to myself and open PRs for both ASAP if this is ok for you?

**Comment by guozhangwang** [comment]
@mjsax #1277 is merged, could you rebase?

**Comment by mjsax** [comment]
@guozhangwang Rebased. Also added class JavaDoc for TimeWindow.

**Comment by guozhangwang** [comment]
Thanks @mjsax for your patience! LGTM and merged to trunk.
